internal/controllers/machine/machine_controller.go

```diff
@@ -1025,3 +1090,193 @@ func (r *Reconciler) nodeToMachine(ctx context.Context, o client.Object) []recon
 
 	return nil
 }
+
+// getAttachedVolumeInformation returns information about volumes attached to the node:
+// * VolumesAttached names from node.Status.VolumesAttached.
+// * PersistentVolume names from VolumeAttachments with status.Attached set to true.
+func getAttachedVolumeInformation(ctx context.Context, remoteClient client.Client, node *corev1.Node) (sets.Set[string], sets.Set[string], error) {
+	attachedVolumeName := sets.Set[string]{}
+	attachedPVNames := sets.Set[string]{}
+
+	for _, attachedVolume := range node.Status.VolumesAttached {
+		attachedVolumeName.Insert(string(attachedVolume.Name))
+	}
+
+	volumeAttachments, err := getVolumeAttachmentForNode(ctx, remoteClient, node.GetName())
+	if err != nil {
+		return nil, nil, err
+	}
+
+	for _, va := range volumeAttachments {
+		// Return an error if a VolumeAttachments does not refer a PersistentVolume.
+		if va.Spec.Source.PersistentVolumeName == nil {
+			return nil, nil, errors.Errorf("PersistentVolumeName for VolumeAttachment %s is nil", va.GetName())
+		}
+		attachedPVNames.Insert(*va.Spec.Source.PersistentVolumeName)
+	}
+
+	return attachedVolumeName, attachedPVNames, nil
+}
+
+// getPersistentVolumeClaimsToIgnore gets all pods which have been ignored by drain and returns a list of
+// NamespacedNames for all PersistentVolumeClaims referred by the pods.
+// Note: this does not require us to list PVC's directly.
+func getPersistentVolumeClaimsToIgnore(ctx context.Context, remoteClient client.Client, nodeName string) (sets.Set[string], error) {
+	drainHelper := drain.Helper{
+		Client: remoteClient,
+	}
+
+	pods, err := drainHelper.GetPodsForEviction(ctx, nodeName)
```

Comment:
+100 to re evaluate the list of pods for which we have to skip volume detach

Reply:
We also don't have a choice. There is no guarantee drainNode was run in the same reconcile